DateTimeUtils.dfy

```diff
@@ -0,0 +1,141 @@
+include "DateTimeConstant.dfy"
+
+module DateTimeUtils {
+  import opened Std.Strings
+  import opened Std.BoundedInts
+  import opened DateTimeConstant
+
+  // Month names for better error messages
+  const MONTH_NAMES: seq<string> := ["January", "February", "March", "April", "May", "June",
+                                     "July", "August", "September", "October", "November", "December"]
+
+  // Function versions for use in function contexts
+  function {:extern "DateTimeImpl", "ToEpochTimeMilliseconds"}
+    {:axiom} ToEpochTimeMillisecondsFunc(year: int32, month: uint8, day: uint8, hour: uint8, minute: uint8, second: uint8, millisecond: uint16): int
+
+  function {:extern "DateTimeImpl", "FromEpochTimeMilliseconds"}
+    {:axiom} FromEpochTimeMillisecondsFunc(epochMillis: int): seq<int32>
+    ensures |FromEpochTimeMillisecondsFunc(epochMillis)| == 7
+    ensures var components := FromEpochTimeMillisecondsFunc(epochMillis);
+            IsValidDateTime(components[0], components[1] as uint8, components[2] as uint8, components[3] as uint8, components[4] as uint8, components[5] as uint8, components[6] as uint16)
+
+  // External function for getting current time components
+  function {:extern "DateTimeImpl", "GetNowComponents"}
+    {:axiom} GetNowComponentsFunc(): seq<int32>
+    ensures |GetNowComponentsFunc()| == 7
+    ensures var components := GetNowComponentsFunc();
+            MIN_YEAR <= components[0] <= MAX_YEAR && // year as int32
+            forall i :: 1 <= i < |components| ==> 0 <= components[i] <= 65535
+
+  // Leap year calculation
+  predicate IsLeapYear(year: int32)
+  {
+    (year % 4 == 0) && (year % 100 != 0 || year % 400 == 0)
+  }
+
+  // Days in month calculation
+  function DaysInMonth(year: int32, month: uint8): uint8
+    requires 1 <= month <= 12
+  {
+    if month == 2 then
+      if IsLeapYear(year) then 29 else 28
+    else if month == 4 || month == 6 || month == 9 || month == 11 then 30
+    else 31
+  }
+
+  // Days in year calculation
+  function DaysInYear(year: int32): int
+  {
+    if IsLeapYear(year) then 366 else 365
+  }
+
+  // Month name getter
+  function GetMonthName(month: uint8): string
+    requires 1 <= month <= 12
+  {
+    MONTH_NAMES[month - 1]
+  }
+
+  // Date-time validation predicate
+  predicate IsValidDateTime(year: int32, month: uint8, day: uint8, hour: uint8, minute: uint8, second: uint8, millisecond: uint16)
+  {
+    MIN_YEAR <= year <= MAX_YEAR &&
+    1 <= month <= 12 &&
+    1 <= day <= (DaysInMonth(year, month) as uint8) &&
```

Comment:
I actually think we should simplify this and just make it `1 <= day <= MAX_DAYS_PER_MONTH` (i.e. 31). That's only slightly weaker, and it means we no longer need to worry about leap years in Dafny code.